feat: Adds event owner role to the system

shreyanshdwivedi opened the pull request:
<!--
(Thanks for sending a pull request! Please make sure you click the link above to view the contribution guidelines, then fill out the blanks below.)
-->
<!-- Add the issue number that is fixed by this PR (In the form Fixes #123) -->

Fixes #6001 

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes <!-- use `nosetests tests/all` to run all the tests -->
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
<!-- If an existing function does not have a docstring, please add one -->
- [ ] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:
Currently, owners of the event cannot be differentiated. So organizer event role is to be changed to owner, to differentiate and make it clear

#### Changes proposed in this pull request:
Adds event `owner` role to the system

shreyanshdwivedi commented:
@iamareebjamal I wanted to ask on thing. How are we storing roles in db? I see 6 roles in db but couldn't find any migration which adds those roles.
I want to update the roles table and insert `owner` at `id=1`. Can you please give me a little background?

CosmicCoder96 commented:
Why do you want to store it at ID 1? We don’t manually set the ID of the
records. Role is searched by its name everywhere not ID

On Thu, 13 Jun 2019 at 5:26 PM, Shreyansh Dwivedi <notifications@github.com>
wrote:

> @iamareebjamal <https://github.com/iamareebjamal> I wanted to ask on
> thing. How are we storing roles in db? I see 6 roles in db but couldn't
> find any migration which adds those roles.
> I want to update the roles table and insert owner at id=1. Can you please
> give me a little background?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/fossasia/open-event-server/pull/6042?email_source=notifications&email_token=AEDUDRN6PGWK44MU3524XZTP2IHFRA5CNFSM4HXWQNC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXTDBMI#issuecomment-501625009>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEDUDRNYHCCVSUBGECSSPC3P2IHFRANCNFSM4HXWQNCQ>
> .
>

shreyanshdwivedi commented:
Actually there are a few tables where id 1 is being used to associate event and role. So I thought I could add a role at id 1 which will automatically make the user the owner. I thought I could do it in a single go.
However now I feel that there might be more than 1 organizer in production and there can be only one owner. So this will not be good.
I will add role and use its id to make 1st organizer the event owner.

CosmicCoder96 commented:
Okay, as to the answer to your original question I believe the roles are
created in the create DB script.

On Thu, 13 Jun 2019 at 5:33 PM, Shreyansh Dwivedi <notifications@github.com>
wrote:

> Actually there are a few tables where id 1 is being used to associate
> event and role. So I thought I could add a role at id 1 which will
> automatically make the user the owner. I thought I could do it in a single
> go.
> However now I feel that there might be more than 1 organizer in production
> and there can be only one owner. So this will not be good.
> I will add role and use its id to make 1st organizer the event owner.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/fossasia/open-event-server/pull/6042?email_source=notifications&email_token=AEDUDRKU5IJHQ6ZVKA5P5O3P2IH5LA5CNFSM4HXWQNC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXTDSLI#issuecomment-501627181>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEDUDRNBRGS2YFBS3LSMIMTP2IH5LANCNFSM4HXWQNCQ>
> .
>

iamareebjamal commented:
> I believe the roles are created in the create DB script

This is problematic 

Make a non conflicting migration for it

CosmicCoder96 commented:
@shreyanshdwivedi here,
https://github.com/fossasia/open-event-server/blob/f192ad56f630383a36c3eec47a3140a821968b0f/populate_db.py#L60

shreyanshdwivedi commented:
@CosmicCoder96 yes I got that earlier but how will it reflect in my db because changing it is not affecting my roles table.
I think I'll have to make a migration file to add roles in db as mentioned by @iamareebjamal above. But if the order of the roles will change, their ids will changes and other tables which stores id of roles will be affected.
I'll check what I can do to avoid any conflicts

CosmicCoder96 commented:
@shreyanshdwivedi @iamareebjamal I don't think looking up for roles using their ID is ever a good idea, their name or slug is their distinguishing factor, shouldn't they be referred to using that, in which case this task can be merely reduced to adding the owner role in the populate script, and writing a migration which adds the role of owner, IF no other role with the name owner exists.  

@shreyanshdwivedi Can you point me to where 

> Actually there are a few tables where id 1 is being used to associate event and role

I think they should be replaced with 
```
role = Role.query.filter_by(name=ORGANIZER).first()
```
Like it has been used at some places in the app. (This example is from the events API).

shreyanshdwivedi commented:
@CosmicCoder96 you can refer to `users_events_roles` table in db for instance where `event_id`, `user_id` and `role_id` are the columns.

CosmicCoder96 commented:
@shreyanshdwivedi For the legacy data, for each event, in the user events table the record which was inserted first (with an organizer relationship), denotes the creator of the event. You can leave the original data as is, create a migration to add owner role if it does not exist in roles table, and also add it to the end of populate db file. Next, add a migration to find the implicit creator of the event, (using the above logic), find the id of the entry in roles table whose name is 'OWNER' and insert into user_events_roles table the user_id, role_id(for the owner, whatever it may be ) and the event ID. What do you think?

shreyanshdwivedi commented:
I totally agree to the steps you suggested. Just one thing, when we find out the owner of the event, I'll suggest to update the role_id of the entry in users_events_roles table rather than inserting a new one as it may lead to confusion later. If we update the role_id only, it'll be better. Does it sound correct?

CosmicCoder96 commented:
But an event owner is still an organiser yes ? Or no?

On Fri, 14 Jun 2019 at 2:52 AM, Shreyansh Dwivedi <notifications@github.com>
wrote:

> I totally agree to the steps you suggested. Just one thing, when we find
> out the owner of the event, I'll suggest to update the role_id of the entry
> in users_events_roles table rather than inserting a new one as it may lead
> to confusion later. If we update the role_id only, it'll be better. Does it
> sound correct?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/fossasia/open-event-server/pull/6042?email_source=notifications&email_token=AEDUDRIQTLGT2SYULWDHXKTP2KJO5A5CNFSM4HXWQNC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXUWH3Q#issuecomment-501834734>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEDUDROH6J4LKNOENLR77P3P2KJO5ANCNFSM4HXWQNCQ>
> .
>

shreyanshdwivedi commented:
@CosmicCoder96 yes. An event owner will have all the rights of an organizer just like an organizer has all the rights of a coorganizer but the roles will still be different. Later we will provide more rights to the owner

bhaveshAn commented:
@shreyanshdwivedi Areeb is right. Migration should also be taken care of if modifying the ID. Else if we create owner role after all the pre existing roles in `create_roles` that's also not an issue. As there is no hard and fast rule that the roles should be created in decreasing order of their degree.

shreyanshdwivedi commented:
@CosmicCoder96 @iamareebjamal @bhaveshAn I've added a migration file that renames the column name too and introduce a new role in server. It also updates the first organizer to owner. 
The codacy is failing because of argument named `f` but there are other functions in same file which follow same thing. That's why I followed this convention.
Please review

shreyanshdwivedi commented:
@iamareebjamal @CosmicCoder96 @niranjan94 please review

CosmicCoder96 commented on app/api/__init__.py:
```diff
@@ -315,6 +315,8 @@
           '/events/<identifier>/relationships/stripe-authorization')
 # Events -> roles:
+api.route(EventRelationship, 'event_owner', '/events/<int:id>/relationships/owner',
+          '/events/<identifier>/relationships/owner')
```
what about the corresponding user relationship?

shreyanshdwivedi replied:
@CosmicCoder96 I've added owner to event relationship -
`api.route(UserRelationship, 'user_owner_event', '/users/<int:id>/relationships/owner-events')`

CosmicCoder96 commented on app/api/schema/events.py:
```diff
@@ -283,6 +283,13 @@ def validate_timezone(self, data, original_data):
+    owner = Relationship(attribute='owner',
+                         self_view='v1.event_owner',
+                         self_view_kwargs={'id': '<id>'},
+                         related_view='v1.user_list',
```
how can an event have multiple owners, wasn't the whole point of the owner was to differentiate between creator and organizers?

shreyanshdwivedi replied:
@CosmicCoder96 missed it. Fixed now

CosmicCoder96 commented on app/models/event.py:
```diff
@@ -125,6 +125,13 @@ class Event(SoftDeletionModel):
         'EventSubTopic', backref='event', foreign_keys=[event_sub_topic_id])
     events_orga = db.relationship(
         'EventOrgaModel', backref='event', foreign_keys=[events_orga_id])
+    owner = db.relationship('User',
```
this will return an array/list with a single user inside it, check docs of alchemy to restrict it to a single result.

shreyanshdwivedi replied:
@CosmicCoder96 I've used - `owner = db.relationship('User', backref='owner_events')` now

shreyanshdwivedi commented:
@CosmicCoder96 @iamareebjamal I've updated the PR and made required changes. Please review

mrsaicharan1 commented on app/api/helpers/permissions.py:
```diff
@@ -103,6 +103,27 @@ def decorated_function(*args, **kwargs):
+    def decorated_function(*args, **kwargs):
+        user = current_identity
+
+        if user.is_staff:
```
Do we need this as this is only for an owner?

shreyanshdwivedi replied:
Actually this is needed. It gives admin, superadmin access to as owner of the event. An admin and superadmin should have all the rights. That's why included this like other permissions

uds5501 commented on app/api/events.py:
```diff
@@ -154,19 +154,26 @@ def before_post(self, args, kwargs, data=None):
         """
-        role = Role.query.filter_by(name=ORGANIZER).first()
         user = User.query.filter_by(id=view_kwargs['user_id']).first()
+        role = Role.query.filter_by(name=OWNER).first()
```
@shreyanshdwivedi after event creation, shouldn't there be a role for `organiser` too for the owner?

shreyanshdwivedi replied:
@uds5501 when event is created the user creating the event will be the owner. 
He can then send role-invites to user to become organizer.

iamareebjamal replied:
Will the owner have same/more permissions than an organizer? If not, fix it

shreyanshdwivedi replied:
For now owner will have same permissions as that of an organizer. Later we can give more permissions to owner

uds5501 commented on app/api/helpers/export_helpers.py:
```diff
@@ -29,8 +29,8 @@
         'id', 'name', 'latitude', 'longitude', 'location_name', 'starts_at', 'ends_at',
-        'timezone', 'description', 'original_image_url', 'logo_url', 'organizer_name',
-        'organizer_description', 'external_event_url', 'ticket_url', 'privacy', 'event_type_id',
+        'timezone', 'description', 'original_image_url', 'logo_url', 'owner_name',
```
@shreyanshdwivedi IMO, organizer name should be preserved too for the sake of upcoming roles or we are using only owner for now?

shreyanshdwivedi replied:
@uds5501 we can get details of organizers via Relationship we have. Only owner info should suffice here as it can be used for filtering too.

shreyanshdwivedi commented:
@iamareebjamal @mrsaicharan1 @uds5501  few updates - 
I ran through the whole system while I was on this branch. I successfully -
 - Created event
 - Order ticket
 - Transferred event
 - Invited users to different roles
 - Created and Updated sessions
 - Used postman to check the routes I introduced. 
I encountered 2 bugs that were there for a long time and reported them.
Please review this PR so that we can move to other linked PRs. Thank you

uds5501 commented:
@shreyanshdwivedi Understood. I am pulling the changes and testing in my system.

iamareebjamal commented:
Other than the changes requested, looks good

iamareebjamal commented on app/api/events.py:
```diff
@@ -154,19 +154,26 @@ def before_post(self, args, kwargs, data=None):
+        # save_to_db(uer, 'Event Saved')
+        # role_invite = RoleInvite(user.email, role.title_name, event.id, role.id, datetime.now(pytz.utc),
+        #                          status='accepted')
+        # save_to_db(role_invite, 'Organizer Role Invite Added')
```
Remove commented lines

iamareebjamal commented on app/api/helpers/permission_manager.py:
```diff
@@ -94,7 +109,8 @@ def is_coorganizer_endpoint_related_to_event(view, view_args, view_kwargs, *args
 
-    if user.is_organizer(kwargs['event_id']) or user.is_coorganizer(kwargs['event_id']):
+    if user.is_owner(kwargs['event_id']) or user.is_organizer(kwargs['event_id']) or \
+            user.is_coorganizer(kwargs['event_id']):
```
These 3 conditions are very common and repeated. Move to user model in a function and repeat everywhere

shreyanshdwivedi commented:
> @shreyanshdwivedi Understood. I am pulling the changes and testing in my system.

@uds5501 do let us know, please. This is a big refactor. Your input is crucial and important to us
Thank You.

uds5501 commented on app/api/events.py:
```diff
@@ -46,7 +46,7 @@
 from app.models.track import Track
 from app.models.user_favourite_event import UserFavouriteEvent
-from app.models.user import User, ATTENDEE, ORGANIZER, COORGANIZER
+from app.models.user import User, ATTENDEE, OWNER, ORGANIZER, COORGANIZER
```
@shreyanshdwivedi The basics of event an seems to be working fine. Although you might want to create a follow up PR on the front end regarding this (for existing and new events):
![image](https://user-images.githubusercontent.com/21087061/60417177-5fee6180-9bfd-11e9-97d2-2a8fc46e2bbe.png)

shreyanshdwivedi replied:
@uds5501 Yeah, a follow-up PR is required to change few more things on frontend. I'll do it once this PR is merged

uds5501 replied:
@shreyanshdwivedi Apologies for not elaborating.
There should be a role of "owner" displayed on the dashboard and the manage event table

shreyanshdwivedi replied:
Sure @uds5501. Will open a followup PR on FE once this is merged. Thanks for the input

shreyanshdwivedi commented:
@uds5501 @iamareebjamal plz review

iamareebjamal commented on app/api/helpers/permission_manager.py:
```diff
@@ -176,7 +189,7 @@ def is_speaker_itself_or_admin(view, view_args, view_kwargs, *args, **kwargs):
         return view(*view_args, **view_kwargs)
 
-    if user.is_organizer(kwargs['event_id']) or user.is_coorganizer(kwargs['event_id']):
+    if user.is_event_head(kwargs['event_id']):
```
has_event_access

shreyanshdwivedi commented:
@iamareebjamal did the changes. Please review

shreyanshdwivedi commented:
@CosmicCoder96 please review

iamareebjamal commented:
Fix Travis Build

iamareebjamal commented:
Most probably due to split migration heads, and the script did not fail correctly

shreyanshdwivedi commented:
Yeah I'll have to update the migration to recent merged one
`Error: Multiple heads are present; please specify a single target revision` is in travis failure

shreyanshdwivedi commented:
@CosmicCoder96 @iamareebjamal please review
